---
title: "Bedtools Merge of m6A Peaks"
output: rmdformats::readthedown
author: "Christina Fitzsimmons"
date: "Last updated 2021-02-08"
---

# Introduction
Using this notebook to determine which `bedtools merge` options are the best for our data. 
```{r packages}
library(tidyverse)
library(purrr)
```
# 1. Analysis of various bedtools merge parameters

```{r Generate 100bp Windows}

library(tidyverse)
library(purrr)

# Need to play with purrr a bit more. 

WT1 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT1_summits.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "summit_start"=V2, "summit_end"=V3, "name"=V4, "score"=V5) %>%
  dplyr::mutate("end" = summit_start+50, "start"=summit_start-50) %>%
  dplyr::select(chr, start, end, name, score) %>%
  dplyr::filter(grepl('chr', chr))

WT2 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT2_summits.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "summit_start"=V2, "summit_end"=V3, "name"=V4, "score"=V5) %>%
  dplyr::mutate("end" = summit_start+50, "start"=summit_start-50) %>%
  dplyr::select(chr, start, end, name, score) %>%
  dplyr::filter(grepl('chr', chr))

WT3 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT3_summits.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "summit_start"=V2, "summit_end"=V3, "name"=V4, "score"=V5) %>%
  dplyr::mutate("end" = summit_start+50, "start"=summit_start-50) %>%
  dplyr::select(chr, start, end, name, score) %>%
  dplyr::filter(grepl('chr', chr))

MUT1 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/MUT1_summits.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "summit_start"=V2, "summit_end"=V3, "name"=V4, "score"=V5) %>%
  dplyr::mutate("end" = summit_start+50, "start"=summit_start-50) %>%
  dplyr::select(chr, start, end, name, score) %>%
  dplyr::filter(grepl('chr', chr))

MUT2 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/MUT2_summits.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "summit_start"=V2, "summit_end"=V3, "name"=V4, "score"=V5) %>%
  dplyr::mutate("end" = summit_start+50, "start"=summit_start-50) %>%
  dplyr::select(chr, start, end, name, score) %>%
  dplyr::filter(grepl('chr', chr))

MUT3 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/MUT3_summits.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "summit_start"=V2, "summit_end"=V3, "name"=V4, "score"=V5) %>%
  dplyr::mutate("end" = summit_start+50, "start"=summit_start-50) %>%
  dplyr::select(chr, start, end, name, score) %>%
  dplyr::filter(grepl('chr', chr))

# First write out all files individually, then write out combined files of all WT and all Mutant
write.table(WT1, file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT1_100bp_windows.bed", row.names = FALSE, col.names = FALSE, sep = "\t", quote = FALSE)

write.table(WT2, file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT2_100bp_windows.bed", row.names = FALSE, col.names = FALSE, sep = "\t", quote = FALSE)

write.table(WT3, file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT3_100bp_windows.bed", row.names = FALSE, col.names = FALSE, sep = "\t", quote = FALSE)

write.table(MUT1, file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/MUT1_100bp_windows.bed", row.names = FALSE, col.names = FALSE, sep = "\t", quote = FALSE)

write.table(MUT2, file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/MUT2_100bp_windows.bed", row.names = FALSE, col.names = FALSE, sep = "\t", quote = FALSE)

write.table(MUT3, file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/MUT3_100bp_windows.bed", row.names = FALSE, col.names = FALSE, sep = "\t", quote = FALSE)


#files = dir(path = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/", pattern = "*.bed")
#files

#filenames <- substr(files, 1,6)
#for(i in filenames) {
#  file.path(files, paste(i, ".bed", sep=""))
#  assign(i, read.delim(filepath, sep = "\t"))
#}
#test <- generate_windows(files)

## 1.1 Analysis of various overlap integers for the d parameter
By default, only overlapping or book-ended features are combined into a new feature. However, one can force merge to combine more distant features with the -d option. For example, were one to set -d to 1000, any features that overlap or are within 1000 base pairs of one another will be combined.

From the help menu
- Note: negative values enforce the number of b.p. required for overlap (maybe we want this??)
```{bash bedtools merge, eval=FALSE}
module load bedtools/2.30.0

# original merge 
bedtools merge -c 4 -d 10 -o count,collapse -i WTall.sorted.100bp.windows.bed > mergedWT_d10.bed
bedtools merge -c 4 -d 10 -o count,collapse -i MUTall.sorted.curated.100bp_v2.bed > mergedMUT_d10.bed

# merge 25
bedtools merge -c 4 -d 25 -o count,collapse -i WTall.sorted.100bp.windows.bed > mergedWT_d25.bed
bedtools merge -c 4 -d 25 -o count,collapse -i MUTall.sorted.curated.100bp_v2.bed > mergedMUT_d25.bed

# merge 50 
bedtools merge -c 4 -d 50 -o count,collapse -i MUTall.sorted.curated.100bp_v2.bed > mergedMUT_d50.bed
bedtools merge -c 4 -d 50 -o count,collapse -i WTall.sorted.100bp.windows.bed > mergedWT_d50.bed

# merge negative 10
bedtools merge -c 4 -d -10 -o count,collapse -i WTall.sorted.100bp.windows.bed > mergedWT_minus10.bed
bedtools merge -c 4 -d -10 -o count,collapse -i MUTall.sorted.curated.100bp_v2.bed > mergedMUT_minus10.bed

```
#2. Histograms of the data

## 2.1 Reading in files
```{r read in merged bedfiles}

# Wildtype Bedfiles
WT_d10 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedWT_d10.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)

WT_d25 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedWT_d25.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)

WT_d50 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedWT_d50.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)

WT_minus10 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedWT_minus10.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)

WT_minus25 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedWT_minus25.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)


WT_stats <- WT_minus10 %>% dplyr::filter(counts == 2 | counts == 3)
nrow(WT_stats)


# Mutant Bedfiles
Mut_d10 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedMUT_d10.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)

Mut_d25 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedMUT_d25.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)

Mut_d50 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedMUT_d50.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)

Mut_minus10 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/mergedMUT_minus10.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr" = V1, "start"=V2, "end"=V3, "counts"=V4, "filenames"=V5)
```
## 2.2 Generating histograms

```{r plotting histograms}

WTd10_plot <- ggplot(WT_d10, aes(x = counts)) +
  geom_histogram() +
  scale_y_log10() +
  theme_bw() +
  labs(title="Bedtools Merge", subtitle = "WT d10", x = "Number of overlapping windows", y= "Log10 counts")
WTd10_plot

WTd25_plot <- ggplot(WT_d25, aes(x = counts)) +
  geom_histogram() +
  scale_y_log10() +
  theme_bw() +
  labs(title="Bedtools Merge", subtitle = "WT d25", x = "Number of overlapping windows", y= "Log10 counts")
WTd25_plot

WTd50_plot <- ggplot(WT_d50, aes(x = counts)) +
  geom_histogram() +
  scale_y_log10() +
  theme_bw() +
  labs(title="Bedtools Merge", subtitle = "WT d50", x = "Number of overlapping windows", y= "Log10 counts")
WTd50_plot



# Negative D-value plots

WTminusd10_plot <- ggplot(WT_minus10, aes(x = counts)) +
  geom_histogram() +
  scale_y_log10() +
  theme_bw() +
  labs(title="Bedtools Merge", subtitle = "WT minus 10", x = "Number of overlapping windows", y= "Log10 counts")
WTminusd10_plot

WTminusd25_plot <- ggplot(WT_minus25, aes(x = counts)) +
  geom_histogram() +
  scale_y_log10() +
  theme_bw() +
  labs(title="Bedtools Merge", subtitle = "WT minus 25", x = "Number of overlapping windows", y= "Log10 counts")
WTminusd25_plot
```

## create a set in R? 


```{r}
ALLmerge_minus10 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/ALLmerge_minus10_strandV2.txt", header = TRUE, sep = "\t") 

ALLmerge_minus25 <-read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/ALLmerge_minus25_strandV2.txt", header = TRUE, sep = "\t") 

ALLminusd10_plot <- ggplot(ALLmerge_minus10, aes(x = counts)) +
  geom_histogram() +
  scale_y_log10() +
  theme_bw() +
  labs(title="Bedtools Merge", subtitle = "ALL minus 10", x = "Number of overlapping windows", y= "Log10 counts")
ALLminusd10_plot

ALLminusd25_plot <- ggplot(ALLmerge_minus25, aes(x = counts)) +
  geom_histogram() +
  scale_y_log10() +
  theme_bw() +
  labs(title="Bedtools Merge", subtitle = "ALL minus 25", x = "Number of overlapping windows", y= "Log10 counts")
ALLminusd25_plot

```
```{r}
ALLmerge_minus10_count1 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/ALLmerged_minus10_1count.txt", header = TRUE, sep = "\t") %>%
  dplyr::filter(genotype=='WT')

ALLmerge_minus25_count1 <-read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/ALLmerged_minus25_1count.txt", header = TRUE, sep = "\t") %>%
  dplyr::filter(genotype=='WT')
```
Bedtools intersect Questions
```{r}

WT_intersect_original <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT_2or3count_v35-longest_intersect.txt", header = FALSE, sep = "\t")

WT_intersect_wao <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT_2or3count_gencodeV35_intersection_v2.txt", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr"=V1, "start"=V2, "end"=V3, "gene"=V7, "strand"=V9, "overlap"=V10) %>%
  dplyr::filter(overlap > 0)

plot1 <- ggplot(WT_intersect_wao, aes(x = overlap)) +
  geom_histogram() +
  theme_bw() +
  labs(title="Bedtools Intersect", subtitle = "No overlap restrictions", x = "Number of bases overlapping with gencodeV35 ", y= "counts")
plot1


WT_intersect_f25 <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT_intersect_f0.25.txt", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr"=V1, "start"=V2, "end"=V3, "gene"=V7, "strand"=V9, "overlap"=V10) %>%
  dplyr::filter(overlap > 0)

WT_intersect_f50<- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/WT_intersect_f0.50.txt", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr"=V1, "start"=V2, "end"=V3, "gene"=V7, "strand"=V9, "overlap"=V10) %>%
  dplyr::filter(overlap == 0)


plot2 <- ggplot(WT_intersect_f25, aes(x = overlap)) +
  geom_histogram() +
  theme_bw() +
  labs(title="Bedtools Intersect", subtitle = "Fraction Overlap = 25%", x = "Number of bases overlapping with gencodeV35 ", y= "counts")
plot2


plot3 <- ggplot(WT_intersect_f50, aes(x = overlap)) +
  geom_histogram() +
  theme_bw() +
  labs(title="Bedtools Intersect", subtitle = "Fraction Overlap = 50%", x = "Number of bases overlapping with gencodeV35 ", y= "counts")
plot3


```

## What do the merged files look like? Size of window?

```{r}

library(tidyverse)

WT_merged <- read.delim(file = "/Users/fitzsimmonscm/Documents/Projects_and_Data/Batista_Lab_FH_SDHB_Project/Rprojects/UOK_manuscript/m6A_IP/MACS2_output_beds_2021.02/bedtools_merge/WTpeaks_sorted_peakname_2or3count.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename (chromosome=V1, start=V2, end=V3, peakname=V4) %>%
  dplyr::mutate ("window" = end-start)

MUT_merged <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/bedtools_intersect_longest_isoform/MutantPeaks_sorted_peakname_2or3count.bed", header = FALSE, sep = "\t") %>%
  dplyr::rename (chromosome=V1, start=V2, end=V3, peakname=V4) %>%
  dplyr::mutate ("window" = end-start)

quantile(WT_merged$window)
quantile(MUT_merged$window)


plot4 <- ggplot(WT_merged, aes(x = window)) +
  geom_histogram() +
  theme_bw() +
  labs(title="Histogram of WT Peaks After Merging", x = "Size of Window ", y= "counts")
plot4


plot5 <- ggplot(MUT_merged, aes(x = window)) +
  geom_histogram() +
  theme_bw() +
  labs(title="Histogram of Mutant Peaks After Merging", x = "Size of Window ", y= "counts")
plot5


```

# 4. Analysis of the histograms of different intersections
We want to ask the question of how much are things overlapping? 
What kind of numbers are we getting? 

```{r}

WT_intersect_wao <- read.delim(file = "/Users/fitzsimmonscm/Desktop//WT_2or3count_gencodeV35_intersection_v2.txt", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr"=V1, "start"=V2, "end"=V3, "gene"=V7, "score" = V8, "strand"=V9, "overlap"=V10) %>%
  dplyr::filter(overlap == 0)

df1 <- WT_intersect_wao %>%
  dplyr::select(chr, start, end, gene, score, strand)
write_delim(df1, "/Users/fitzsimmonscm/Desktop/2021.02.17_WT_intersect.bed", col_names = FALSE, delim = "\t")

WT_intersect_counts <- WT_intersect_wao %>%
  group_by(chr, start, end) %>%
  add_count() %>%
  dplyr::rename (multimap = n) %>%
  dplyr::filter (multimap == 2) %>%
  add_count(strand, name = "strand_count")

WT_same <- WT_intersect_counts %>%
  dplyr::filter (strand_count == 1)

# write.csv(WT_intersect_counts, "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/2021.02.17_WTintersection_multimap2.csv", row.names = FALSE)
  

plot6 <- ggplot(WT_same, aes(x = overlap)) +
  geom_histogram() +
  theme_bw() +
  labs(title="Histogram of WT peaks that intersect 2 features on opposite strands ", x = "Size of overlap window (bp) ", y= "counts")
plot6



### Intersection and plots of the mutant dataframe


Mutant_intersect_wao <- read.delim(file = "/Users/fitzsimmonscm/Desktop/MACS2_output_beds_2021.02/Mut_2or3count_gencodeV35_intersection_v2.txt", header = FALSE, sep = "\t") %>%
  dplyr::rename("chr"=V1, "start"=V2, "end"=V3, "peakID" = V4, "gene"=V8,"score" =V9, "strand"=V10, "overlap"=V11)

#df2 <- Mutant_intersect_wao %>%
#  dplyr::select(chr, start, end, gene, score, strand)
#write_delim(df2, "/Users/fitzsimmonscm/Desktop/2021.02.17_MUT_intersect.bed", col_names = FALSE, delim = "\t")

# This is the list for the mutant counts
Mutant_intersect_counts <- Mutant_intersect_wao %>%
  group_by(chr, start, end) %>%
  add_count() %>%
  dplyr::rename (multimap = n) %>%
  dplyr::filter (multimap == 2) %>%
  add_count(strand, name = "strand_count")

Mutant_same <- Mutant_intersect_counts %>%
  dplyr::filter (strand_count == 1)


# write.csv(Mutant_intersect_counts, "/Users/fitzsimmonscm/Desktop/2021.02.21_Mutant_intersect_overlap2.csv", row.names = FALSE)


plot7  <- ggplot(Mutant_same, aes(x = overlap)) +
  geom_histogram() +
  theme_bw() +
  labs(title="Histogram of Mutant peaks that intersect 2 features on opposite strands ", x = "Size of overlap window (bp) ", y= "counts")
plot7

quantile(WT_intersect_counts$overlap)
quantile(Mutant_intersect_counts$overlap)
```

# Generating peak files for counting
Pedro and I have come up with an idea for how best to filter the peak files. At the moment, we're going to retain everything, but mark the peaks in the score column so that we know which peak comes from which type of filter (eg matches 1 feature, matches 2 features same strand, etc). I need to repeat the WT counting with the peaknames in place of the chr /start /stop filtering I have right now. 


```{r Generating the mutant peaklist}

library(tidyverse)


Mutant_intersect_wao <- read.delim(file = "/Users/fitzsimmonscm/Desktop/Mut_2or3count_gencodeV35_intersection_v2.txt", header = FALSE, sep = "\t") %>% dplyr::rename("chr"=V1, "start"=V2, "end"=V3, "peakID" = V4, "gene"=V8,"score" =V9, "strand"=V10, "overlap"=V11)

# Generating a new peak list 
Mutant_peaklist <- Mutant_intersect_wao %>%
  dplyr::group_by(peakID) %>%
  add_count(name = 'peakcounts')




```

